### PR TITLE
Modify product canonical url and title value

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Modify product canonical url and title value for SEO purposes, the title contains the combination attribute infos, the canonical url is the default combination
| Type?         | improvement
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/9508
| How to test?  | In the product page source code check the canonical tag and the title page, title has the current page attributes, canonical is the default combination Try changing the default combination in the admin and check the canonical has changed

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9456)

### CODE REVIEW

File: controllers/front/ProductController.php [hunk 1]
```diff
@@ -1228,6 +1228,18 @@ public function getTemplateVarPage()
         }
         $page['admin_notifications'] = array_merge($page['admin_notifications'], $this->adminNotifications);
 
+        $idProductAttribute = $this->getIdProductAttribute();
+        if ($idProductAttribute) {
+            $attributes = $this->product->getAttributeCombinationsById($this->getIdProductAttribute(), $this->context->language->id);
+            if ($attributes && count($attributes) > 0) {
+                $pageTitle = $page['meta']['title'];
+                foreach ($attributes as $attribute) {
+                    $pageTitle .= ' '.$attribute['group_name'].' '.$attribute['attribute_name'];
```
Comment (PierreRambaud) [inline controllers/front/ProductController.php]: Can directly concat to ` $page['meta']['title']`

File: src/Adapter/Presenter/Product/ProductLazyArray.php [hunk 2]
```diff
@@ -691,7 +689,7 @@ private function getProductURL(
             $ean13,
             $language->id,
             null,
-            (!$canonical) ? $product['id_product_attribute'] : null,
+            $product['id_product_attribute'],
```
Comment (PierreRambaud) [inline src/Adapter/Presenter/Product/ProductLazyArray.php]: Is this the current product attribute?

File: controllers/front/ProductController.php (hunk 1 above)
Comment (jolelievre) [reply controllers/front/ProductController.php]: done!

File: src/Adapter/Presenter/Product/ProductLazyArray.php (hunk 2 above)
Comment (jolelievre) [reply src/Adapter/Presenter/Product/ProductLazyArray.php]: it is the current variant id depending on the context, it is always filled in the input array before present it (ie instanciating ProductLazyArray)

Comment (PierreRambaud) [reply src/Adapter/Presenter/Product/ProductLazyArray.php]: Should it be the default combination selected in BO?

Comment (jolelievre) [reply src/Adapter/Presenter/Product/ProductLazyArray.php]: you're right! my bad, I thought each combination had their own canonical, but it's the default one for all of them

File: controllers/front/ProductController.php [hunk 3]
```diff
@@ -1228,6 +1228,16 @@ public function getTemplateVarPage()
         }
         $page['admin_notifications'] = array_merge($page['admin_notifications'], $this->adminNotifications);
 
+        $idProductAttribute = $this->getIdProductAttribute(true);
```
Comment (mickaelandrieu) [inline controllers/front/ProductController.php]: can you please create a new function with a meaningful name and don't use a boolean flag?

File: controllers/front/ProductController.php [hunk 4]
```diff
@@ -1228,6 +1228,16 @@ public function getTemplateVarPage()
...
+        if ($idProductAttribute) {
+            $attributes = $this->product->getAttributeCombinationsById($this->getIdProductAttribute(), $this->context->language->id);
+            if ($attributes && count($attributes) > 0) {
```
Comment (mickaelandrieu) [inline controllers/front/ProductController.php]:
```diff
- if ($ttributes
+ if (is_array($attributes)
```

**Review by mickaelandrieu — DISMISSED** [review]
Some code smells, minor fixes to address.

Good job @jolelievre !

File: src/Adapter/Presenter/Product/ProductLazyArray.php (hunk 2 above)
Comment (jolelievre) [reply src/Adapter/Presenter/Product/ProductLazyArray.php]: fixed!

File: controllers/front/ProductController.php (hunk 3 above)
Comment (jolelievre) [reply controllers/front/ProductController.php]: i refactorized this method so there are two methods getIdProductAttributeByRequest and getIdProductAttributeByRequestOrGroup

File: controllers/front/ProductController.php (hunk 4 above)
Comment (jolelievre) [reply controllers/front/ProductController.php]: done!

**Comment by jolelievre** [comment]
@mickaelandrieu modifications done on private methods

File: controllers/front/ProductController.php [hunk 5]
```diff
@@ -944,49 +944,89 @@ public function getCategory()
...
+     *
+     * @return int|null
+     * @throws PrestaShopException
+     */
+    private function getIdProductAttributeByGroup()
+    {
+        $idProductAttribute = null;
+
+        $groups = Tools::getValue('group');
+        if (!empty($groups)) {
```
Comment (PierreRambaud) [inline controllers/front/ProductController.php]: No need variable here

```php
if (!empty($groups)) {
  return null;
}

return Product::getIdProductAttributeByIdAttributes(
    $this->product->id,
    $groups,
     true
);
```

Comment (jolelievre) [reply controllers/front/ProductController.php]: rather
`if (empty($groups)) {`
but ok I change it

Comment (PierreRambaud) [reply controllers/front/ProductController.php]: Yes, wrong copy & paste :P

Comment (jolelievre) [reply controllers/front/ProductController.php]: done!

**Review by mickaelandrieu — DISMISSED** [review]
about the refactoring ok, but you may have to execute the following command: php bin/php-cs-fixer fix

**Comment by jolelievre** [comment]
Done! but isn't the command rather `php ./vendor/bin/php-cs-fixer fix` ?

**Comment by PierreRambaud** [comment]
@jolelievre Yep. Thinking he wrotes it fast :D

**Review by PierreRambaud — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thanks @jolelievre

**Comment by Maofree** [comment]
hi
at this moment on the 1.7.5 the problem of the canonical url is still present
the same problem with the alternate urls



the new solution of each combination has an own url, gives problems on these pages, they are not indexed by google, because the canonical url is not exist and it will be redirected to the default combination.
your fix will use the canonical url as the default combination, it can be ok but for sites that come from prestashop 1.5 or 1.6 will be a problems

I think the best solution is the one used in 1.5. and 1.6 I do not think is one that satisfies many people, also because it adds another ID in the url that for some creates problems with indexing

bye

**Pull request merged**